connectors/migrations/20241212_backfill_github_data_source_folders.ts

```diff
@@ -0,0 +1,236 @@
+  dataSourceConfig: DataSourceConfig
+) {
+  const codeFolderId = getCodeRootInternalId(repositoryId);
+  await upsertDataSourceFolder({
+    dataSourceConfig,
+    folderId: codeFolderId,
+    parents: [codeFolderId, repositoryNodeId],
+    title: "Code",
+    mimeType: getMimeTypeFromGithubContentNodeType("REPO_CODE"),
+  });
+  return codeFolderId;
+}
+
+async function upsertDirectoryFolderNode(
+  directory: GithubCodeDirectory,
+  repositoryId: number,
+  dataSourceConfig: DataSourceConfig
+) {
+  // This already contains IDs for Code and Repository folders
+  const parents = await getGithubCodeDirectoryParentIds(
+    directory.connectorId,
+    directory.internalId,
+    repositoryId
+  );
+  await upsertDataSourceFolder({
+    dataSourceConfig,
+    folderId: directory.internalId,
+    parents: [directory.internalId, ...parents],
+    title: directory.dirName,
+    mimeType: getMimeTypeFromGithubContentNodeType("REPO_CODE_DIR"),
+  });
+}
+
+async function upsertIssueFolderNode(
+  repositoryId: string,
+  repositoryNodeId: string,
+  dataSourceConfig: DataSourceConfig
+) {
+  const issuesFolderId = getIssuesInternalId(repositoryId);
+  await upsertDataSourceFolder({
+    dataSourceConfig,
+    folderId: issuesFolderId,
+    parents: [issuesFolderId, repositoryNodeId],
+    title: "Issues",
+    mimeType: getMimeTypeFromGithubContentNodeType("REPO_ISSUES"),
+  });
+}
+
+async function upsertDiscussionFolderNode(
+  repositoryId: string,
+  repositoryNodeId: string,
+  dataSourceConfig: DataSourceConfig
+) {
+  const discussionsFolderId = getDiscussionsInternalId(repositoryId);
+  await upsertDataSourceFolder({
+    dataSourceConfig,
+    folderId: discussionsFolderId,
+    parents: [discussionsFolderId, repositoryNodeId],
+    title: "Discussions",
+    mimeType: getMimeTypeFromGithubContentNodeType("REPO_DISCUSSIONS"),
+  });
+}
+
+async function repositoryContainsCode(repository: GithubCodeRepository) {
+  const directory = await GithubCodeDirectory.findOne({
+    where: {
+      connectorId: repository.connectorId,
+      repoId: repository.repoId,
+    },
+  });
+  if (directory) {
+    return true;
+  }
+  const file = await GithubCodeFile.findOne({
+    where: {
+      connectorId: repository.connectorId,
+      repoId: repository.repoId,
+    },
+  });
+  return !!file;
+}
+
+async function getUniqueRepositoryIdsWithIssues(connectorId: number) {
+  const repoIds = await GithubIssue.findAll({
+    where: {
+      connectorId: connectorId,
+    },
+    attributes: ["repoId"],
+    group: ["repoId"],
+  });
+  return repoIds.map((repoId) => repoId.repoId);
+}
+
+async function getUniqueRepositoryIdsWithDiscussions(connectorId: number) {
+  const repoIds = await GithubDiscussion.findAll({
+    where: {
+      connectorId: connectorId,
+    },
+    attributes: ["repoId"],
+    group: ["repoId"],
+  });
+  return repoIds.map((repoId) => repoId.repoId);
+}
+
+makeScript({}, async ({ execute }) => {
+  if (execute) {
```

Comment:
nit: you could use the execute to get a dry run that logs everything without actually doing anything

Reply:
I'm not sure I get what you mean, how would you do that ?

Reply:
I would do the if (execute) not at top level but before every upsert, which would ofc imply change everything so you should just leave it :)

Reply:
Ah. I see, yep that could be useful I guess :)